contracts/coordinator/src/contract/execute.rs

```diff
@@ -55,3 +87,278 @@ pub fn set_active_verifier_set(
+        ctx.verifier_code_id,
+        cosmwasm_std::to_json_binary(&voting_verifier_api::msg::InstantiateMsg {
+            governance_address: verifier_msg.governance_address.clone(),
+            service_registry_address: axelar_wasm_std::nonempty::String::try_from(
+                service_registry_address.into_string(),
+            )
+            .change_context(Error::InstantiateVerifier)?,
+            service_name: verifier_msg.service_name.clone(),
+            source_gateway_address: verifier_msg.source_gateway_address.clone(),
+            voting_threshold: verifier_msg.voting_threshold,
+            block_expiry: verifier_msg.block_expiry,
+            confirmation_height: verifier_msg.confirmation_height,
+            source_chain: verifier_msg.source_chain.clone(),
+            rewards_address: verifier_msg.rewards_address.clone(),
+            msg_id_format: verifier_msg.msg_id_format.clone(),
+            address_format: verifier_msg.address_format.clone(),
+        })
+        .change_context(Error::InstantiateVerifier)?,
+        label,
+    )
+}
+
+fn instantiate_prover(
+    ctx: &InstantiateContext,
+    label: String,
+    gateway_address: Addr,
+    service_registry_address: Addr,
+    multisig_address: Addr,
+    verifier_address: Addr,
+    prover_msg: &ProverMsg,
+) -> error_stack::Result<(WasmMsg, Addr), Error> {
+    launch_contract(
+        &ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        ctx.salt.clone(),
+        ctx.prover_code_id,
+        cosmwasm_std::to_json_binary(&multisig_prover_api::msg::InstantiateMsg {
+            admin_address: ctx.info.sender.to_string().clone(),
+            governance_address: prover_msg.governance_address.to_string().clone(),
+            coordinator_address: ctx.env.contract.address.to_string().clone(),
+            gateway_address: gateway_address.to_string().clone(),
+            multisig_address: multisig_address.to_string().clone(),
+            service_registry_address: service_registry_address.to_string().clone(),
+            voting_verifier_address: verifier_address.to_string().clone(),
+            signing_threshold: prover_msg.signing_threshold,
+            service_name: prover_msg.service_name.to_string(),
+            chain_name: prover_msg.chain_name.clone(),
+            verifier_set_diff_threshold: prover_msg.verifier_set_diff_threshold,
+            encoder: prover_msg.encoder,
+            key_type: prover_msg.key_type,
+            domain_separator: prover_msg.domain_separator,
+        })
+        .change_context(Error::InstantiateProver)?,
+        label,
+    )
+}
+
+struct InstantiateContext<'a> {
+    deps: DepsMut<'a>,
+    info: MessageInfo,
+    env: Env,
+    salt: Binary,
+    gateway_code_id: u64,
+    verifier_code_id: u64,
+    prover_code_id: u64,
+}
+
+pub fn instantiate_chain_contracts(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    deployment_name: String,
+    params: &DeploymentParams,
```

Comment:
I don't think this should be a reference, this function should own the params. It would eliminate some cloning and dereferencing